src/include/sof/math/trig.h

```diff
@@ -137,17 +144,193 @@ static inline int16_t sin_fixed_16b(int32_t th_rad_fxp)
  */
 static inline int16_t cos_fixed_16b(int32_t th_rad_fxp)
 {
+	int32_t th_cdc_fxp;
 	int32_t sign;
 	int32_t b_yn;
 	int32_t xn;
-	int32_t th_cdc_fxp;
-	cordic_cfg type = EN_16B_CORDIC_COSINE;
-	/* compute coeff from angles*/
-	cordic_sin_cos(th_rad_fxp, type, &sign, &b_yn, &xn, &th_cdc_fxp);
+
+	cordic_approx(th_rad_fxp, CORDIC_15B_TABLE_SIZE, &sign, &b_yn, &xn, &th_cdc_fxp);
+
 	th_cdc_fxp = sign * xn;
-	/*convert Q1.31 to Q1.15 format*/
-	return sat_int16(Q_SHIFT_RND((sat_int32(Q_SHIFT_LEFT((int64_t)th_cdc_fxp, 30, 31))),
-	31, 15));
+	/*convert Q2.30 to Q1.15 format*/
+	return sat_int16(Q_SHIFT_RND(th_cdc_fxp, 30, 15));
+}
+
+/**
+ * CORDIC-based approximation of complex exponential e^(j*THETA).
+ * computes COS(THETA) + j*SIN(THETA) using CORDIC algorithm
+ * approximation and returns the complex result.
+ * THETA values must be in the range [-2*pi, 2*pi). The cordic
+ * exponential algorithm converges, when the angle is in the
+ * range [-pi/2, pi/2).If an angle is outside of this range,
+ * then a multiple of pi/2 is added or subtracted from the
+ * angle until it is within the range [-pi/2,pi/2).Start
+ * with the angle in the range [-2*pi, 2*pi) and output has
+ * range in [-1.0 to 1.0]
+ * Error (max = 0.000000015832484), THD+N  = -167.082852232808847
+ * +------------------+-----------------+--------+------------+
+ * | thRadFxp	      |cdccexpth        |thRadFxp|   cdccexpth|
+ * +----+-----+-------+----+----+-------+--------+------------+
+ * |WLen| FLen|Signbit|WLen|FLen|Signbit| Qformat| Qformat    |
+ * +----+-----+-------+----+----+-------+--------+------------+
+ * | 32 | 28  |  1    | 32 | 15 |   1	| 4.28	 | 2.30       |
+ * +------------------+-----------------+--------+------------+
+ */
+static inline struct cordic_cmpx cmpx_exp_32b(int32_t th_rad_fxp)
+{
+	int32_t th_cdc_fxp;
+	int32_t sign;
+	int32_t b_yn;
+	int32_t xn;
+	struct cordic_cmpx cexp;
```

Comment:
cexp needs to be passed in by the caller as can be nuked on the callee's stack (even though it's inline - depends on compiler) i.e.
```c
static inline struct cordic_cmpx cmpx_exp_32b(int32_t th_rad_fxp, struct cordic_cmpx *result)
```
Ditto for other functions.

Reply:
Thanks ! I will take it